paper/paper.md

```diff
@@ -198,10 +205,174 @@ solver_stats = trunk(
+```julia
+using FluxNLPModels
+using CUDA, Flux, NLPModels
+using Statistics
+using Flux.Data: DataLoader
+using Flux: onehotbatch, onecold, @epochs
+using Flux.Losses: logitcrossentropy
+using Base: @kwdef
+using MLDatasets
+using JSOSolvers
+```
+
+
+#### Loading Data
+In this section, we will cover the process of loading datasets and defining minibatches for training your model using Flux. Loading and preprocessing data is an essential step in machine learning, as it allows you to train your model on real-world examples.
+
+We will specifically focus on loading the MNIST dataset. We will divide the data into training and testing sets, ensuring that we have separate data for model training and evaluation.
+
+Additionally, we will define minibatches, which are subsets of the dataset that are used during the training process. Minibatches enable efficient training by processing a small batch of examples at a time, instead of the entire dataset. This technique helps in managing memory resources and improving convergence speed.
+To donwnload and load MNIST dataset [@lecun-bouttou-bengio-haffner1998] from MLDataset:
+```julia
+
+function getdata(args; T = Float32) #T for types
+  ENV["DATADEPS_ALWAYS_ACCEPT"] = "true"
+
+  # Loading Dataset	
+  xtrain, ytrain = MLDatasets.MNIST(Tx = T, split = :train)[:]
+  xtest, ytest = MLDatasets.MNIST(Tx = T, split = :test)[:]
+
+  # Reshape Data in order to flatten each image into a linear array
+  xtrain = Flux.flatten(xtrain)
+  xtest = Flux.flatten(xtest)
+
+  # One-hot-encode the labels
+  ytrain, ytest = onehotbatch(ytrain, 0:9), onehotbatch(ytest, 0:9)
+
+  # Create DataLoaders (mini-batch iterators)
+  train_loader = DataLoader((xtrain, ytrain), batchsize = args.batchsize, shuffle = true)
+  test_loader = DataLoader((xtest, ytest), batchsize = args.batchsize)
+  @info "The type is " typeof(xtrain)
+  return train_loader, test_loader
+end
+```
+#### Loss Function 
+We can define any loss function that we need, here we use Flux build-in logitcrossentropy function. 
+
+```julia
+const loss = logitcrossentropy
+```
+We also definethe loss function `loss_and_accuracy`. It expects the following arguments:
+* ADataLoader object.
+* The `build_model` function we defined above.
+* A device object (in case we have a GPU available).
+  
+```julia
+function loss_and_accuracy(data_loader, model, device; T=Float32)
+  acc = T(0)
+  ls = T(0.0f0)
+  num = T(0)
+  for (x, y) in data_loader
+    x, y = device(x), device(y)
+    ŷ = model(x)
+    ls += loss(ŷ, y, agg = sum)
+    acc += sum(onecold(ŷ) .== onecold(y)) ## Decode the output of the model
+    num += size(x)[end]
+  end
+  return ls / num, acc / num
+end
+```
```

Comment:
true, 
in Knet we have accuracy, so in Flux we need this to calculate the accuracy,
I will add one line to show

Reply:
but the accuracy and loss are problem related, here we assume the output are onehot encoding, that is why I think it would be much better to have it as user defined (similar to what Flux.jl did )
what do you think?

Reply:
What does `NLPModels.obj` returns?

For me, `loss_and_accuracy` complicates the example while it could be in FluxNLPModel.
You told be, you want to potentially change the loss function for a FluxNLPModel, does that mean you will have to define a `loss_and_accuracy` for each loss function you use?

If this is the case, I think it is wrong.

Reply:
So **NLPModels.obj** is accuracy on the mini-batch, in ML when we say accuracy it has different meaning (not sure they use it right but..) basically it is the accuracy on the accuracy set (here is the test data ) and their loss is also on the whole datasets nit just the current mini-batch

Reply:
`NLPModels.obj(::FluxNLPModel, x)` should be the loss function on the current training minibatch, this is what the optimization methods from JSOSolvers will use for the training.

Reply:
that is exactly what that is, NLPModels.obj(::FluxNLPModel, x) is  the loss function on the current training minibatch

Reply:
You can compute accuracy on any dataset (wether training or test), but for new user `accuracy` using the test dataset by default (though an optional argument maybe) is fine.

Reply:
yes but I don't think we should handle it in the fluxNLPModel but the user can define their own method, This allow them to use the same method for Flux and FluxNLPModels and lower the methods in the FluxNLPModels

Reply:
> This allow them to use the same method for Flux and FluxNLPModels and lower the methods in the FluxNLPModels

This is not a receivable argument, it downgrades the features of FluxNLPModels.
You want to work only on the FluxNLPModel and not on both `FluxNLPModel` and Flux architecture.

```julia
accuracy(nlp; data_loader=nlp.test_minibatch_iterator) = ...
```
solves your concerns, you can switch the `data_loader` to anything you want.

Reply:
but if they already have define their accuracy method in Flux and want to integrade to FluxNLPModels?

Reply:
Have you an example to share?

Reply:
sure, 
Eventually in my research we want to work on transformers, here is their code https://github.com/chengchingwen/Transformers.jl/tree/master/src

Reply:
https://github.com/chengchingwen/Transformers.jl/blob/bc393c9c7407a0a9ee5094c4b3a19ec12ab62aa6/example/BERT/mnli/train.jl#L36

Reply:
or another one 
 accuracy = mean(test_predictions .== [onecold(y, langs) for y in test_Y])

Reply:
> Eventually in my research we want to work on transformers,

The transformer loss function is basic, and doesn't depend on the dataset nor the neural network.

Reply:
> or another one accuracy = mean(test_predictions .== [onecold(y, langs) for y in test_Y])

Do you have a link?

Reply:
> > or another one accuracy = mean(test_predictions .== [onecold(y, langs) for y in test_Y])
> 
> Do you have a link?

https://github.com/FluxML/model-zoo/blob/31e147faef61e7f26dc2d6124afbac0cfe518b6a/text/lang-detection/model.jl#L132

Reply:
> 

The idea is that we have different accuracy functions, that user define, the simplicity of them is another discussion

Reply:
Yes, but requiring the user to define a function to check the accuracy of a neural network is wrong, especially if a generic method working in most cases can be written.
At worst, the `accuracy` method fails, and in such case the user must write its own accuracy method.

Reply:
I do not agree with that, especially that I do not find any other DNN library to code the accuracy as a method in them 
Flux.jl, pytorch from python. 
I am not seeing the benefit of having it in the model

Reply:
Two benefits:
- the user don't have to implement this function, something like 
```julia
flux_accuracy(x, y) = mean(Flux.onecold(flux_model(x), classes) .== y);
```
should be generic enough to be applied in most cases;
- the example of the JOSS paper is simpler.

Without a `accuracy` method, you can directly define your FluxNLPModel, train it and not check the result of the training, I find it quite strange.

Reply:
added